contracts/gas/GasEstimate.sol

```diff
@@ -0,0 +1,112 @@
+     * param destinationAddress Destination contract address being called
+     * @param executionGasLimit The gas limit to be used for the destination contract execution,
+     *        e.g. pass in 200k if your app consumes needs upto 200k for this contract call
+     * @return gasEstimate The cross-chain gas estimate, in terms of source chain's native gas token that should be forwarded to the gas service.
+     */
+    function estimateGasFee(
+        string calldata destinationChain,
+        string calldata, /* destinationAddress */
+        bytes calldata payload,
+        uint256 executionGasLimit
+    ) external view returns (uint256 gasEstimate) {
+        GasServiceStorage storage slot = _gasServiceStorage();
+        GasInfo storage gasInfo = slot.gasPrices[destinationChain];
+
+        gasEstimate = gasInfo.baseFee + (executionGasLimit * gasInfo.relativeGasPrice);
+
+        // if chain is L2, compute L1 data fee using ethereum gas price info
+        if (gasInfo.l1ToL2BaseFee != 0) {
+            gasEstimate += computeL1ToL2Fee(
+                payload,
+                slot.gasPrices['ethereum'].relativeGasPrice,
+                gasInfo.l1ToL2BaseFee
+            );
+        }
+    }
+
+    /**
+     * @notice Computes the L1 to L2 fee for a contract call on a destination chain.
+     * @param payload The payload of the contract call
+     * @param l1GasPrice The gas price on the source chain
+     * @param l1ToL2BaseFee The base fee for L1 to L2
+     * @return l1DataFee The L1 to L2 data fee
+     */
+    function computeL1ToL2Fee(
+        bytes calldata payload,
+        uint256 l1GasPrice,
+        uint256 l1ToL2BaseFee
+    ) internal pure returns (uint256 l1DataFee) {
+        /* Optimism Bedrock gas model https://docs.optimism.io/stack/transactions/fees#bedrock
+            tx_data_gas = count_zero_bytes(tx_data) * 4 + count_non_zero_bytes(tx_data) * 16
+            tx_total_gas = (tx_data_gas + fixed_overhead) * dynamic_overhead
+            l1_data_fee = tx_total_gas * ethereum_base_fee
+        */
+
+        // We are using upper bounds for gas estimates
+        // Current OP fixed_overhead value is 188, using 200 to account for fluctuations
+        uint256 fixedOverhead = 200;
+        // Current OP dynamic_overhead_multiplier value is 0.684, using 0.7 to encounter for fluctuations
+        uint256 dynamicOverheadMultiplier = 7;
+        uint256 dynamicOverheadDivisor = 10; // 7 : 10 = 0.7
+
+        uint256 zeroBytesCount;
+        uint256 nonZeroBytesCount;
+        for (uint256 i; i < payload.length; ++i) {
+            if (payload[i] == 0) {
+                ++zeroBytesCount;
+            } else {
+                ++nonZeroBytesCount;
+            }
+        }
+
+        uint256 txDataGas = zeroBytesCount * 4 + nonZeroBytesCount * 16;
+        uint256 txTotalGas = ((txDataGas + fixedOverhead) * l1ToL2BaseFee * dynamicOverheadMultiplier) /
```

Comment:
That's what used in their `Bedrock` model. `l1ToL2BaseFee` is used for multiplication by total gas. And `l1GasPrice` would be used to convert between Eth/Ava token price difference. I will switch to `Ecotone` formula anyway

Reply:
Seems like the answer for Optimism here is only to `l1ToL2BaseFee` and not use `l1GasPrice` at all

Reply:
But `relativeGasPrice` is defined to be the following:
```
        uint256 relativeGasPrice; // dest_gas_price * dest_token_market_price / src_token_market_price
```
It is the Eth gas price (i.e base fee + a small tip) multiplied by eth market price / src chain token market price. Multiplying that should be enough instead of adding another notion of `l1tol2basefee`

Reply:
`blob_base_fee` is 0, so we don't need to worry about converting it

Reply:
Will use `l1GasPrice` as agreed